File: identicon/src/main/java/com/lelloman/identicon/classic/ClassicIdenticonHash.java
package com.lelloman.identicon.classic;


/**
 * 	Provides the values for a classic identicon using an int
 */
public class ClassicIdenticonHash {

	public final int value;

	public static int positiveMod(long i, int m) {
		return Math.abs((int) (i % m));
	}

	private int[] values;

	public ClassicIdenticonHash(int v){
		value = v;

		values = new int[11];

		// middle tile
		values[0] = positiveMod(value,8) * 4;

		// corners
		values[1] = positiveMod((value >> 3), 32);

		// sides
		values[2] = positiveMod((value >> 8),32);

		// invert middle
		values[3] = positiveMod((value >> 13),2);

		// invert corners
		values[4] = positiveMod((value >> 14),2);

		// invert sides
		values[5] = positiveMod((value >> 15),2);

		// r
		values[6] = positiveMod((value >> 16),16);

		// g
		values[7] = positiveMod((value >> 20),16);

		// b
		values[8] = positiveMod((value >> 24),16);

		// corners rotation
		values[9] = positiveMod((value >> 28), 4);

		// sides rotation
		values[10] = positiveMod((value >> 30), 4);

	}

	public int getMiddle() {
		return values[0];
	}

	public int getCorners() {
		return values[1];
	}

	public int getSides() {
		return values[2];
	}

	public int getInvertMiddle() {
		return values[3];
	}

	public int getInvertCorners() {
		return values[4];
	}

	public int getInvertSides() {
		return values[5];
	}

	public int getR() {
		return values[6];
	}

	public int getG() {
		return values[7];
	}

	public int getB() {
		return values[8];
	}

	public int getCornersRotation() {
		return values[9];
	}

	public int getSidesRotation() {
		return values[10];
	}
}

File: identicon/src/main/java/com/lelloman/identicon/drawable/ClassicIdenticonHash.kt
package com.lelloman.identicon.drawable

import com.lelloman.identicon.util.extractBits

/**
 * Provides the values for a classic identicon using MD5 bytes.
 */
class ClassicIdenticonHash(md5: ByteArray) {

    val middle: Int
    val corners: Int
    val sides: Int
    val invertMiddle: Int
    val invertCorners: Int
    val invertSides: Int
    val r: Int
    val g: Int
    val b: Int
    val cornersRotation: Int
    val sidesRotation: Int
    val middleRotation: Int

    init {
        var offset = 0

        // middle tile: 4 bits, mod 9 → index into symmetric tiles
        middle = extractBits(md5, offset, 4) % 9
        offset += 4

        // corners tile: 6 bits, mod 37
        corners = extractBits(md5, offset, 6) % 37
        offset += 6

        // sides tile: 6 bits, mod 37
        sides = extractBits(md5, offset, 6) % 37
        offset += 6

        // invert flags: 1 bit each
        invertMiddle = extractBits(md5, offset, 1)
        offset += 1
        invertCorners = extractBits(md5, offset, 1)
        offset += 1
        invertSides = extractBits(md5, offset, 1)
        offset += 1

        // RGB: 6 bits each, range 0-63
        r = extractBits(md5, offset, 6) % 64
        offset += 6
        g = extractBits(md5, offset, 6) % 64
        offset += 6
        b = extractBits(md5, offset, 6) % 64
        offset += 6

        // rotations: 2 bits each
        cornersRotation = extractBits(md5, offset, 2)
        offset += 2
        sidesRotation = extractBits(md5, offset, 2)
        offset += 2
        middleRotation = extractBits(md5, offset, 2)
    }
}
